### PR TITLE
First stab at concurrent execution

### PR DESCRIPTION
**Pull request by deosjr**
* added naive or-parallellism
* example given using exclusive disjunction
* example given using normal disjunction, harder to gauge effectiveness
* two more examples of detecting places where parallelism could help in unroll.go
* and-parallellism is way less trivial

### CODE REVIEW

File: mini/unroll_test.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			t.Fatalf("expected %d, but got %d instead", i, *sexpr.Atom.Int)
+		}
+	}
+}
+
+func TestMapO(t *testing.T) {
+	list := ast.Cons(ast.NewInt(0), ast.Cons(ast.NewInt(1), ast.Cons(ast.NewInt(2), nil)))
+	sexprs := micro.Run(-1, func(q *ast.SExpr) micro.Goal {
+		return MapO(
+			micro.EqualO,
```
Comment (awalterschulze) [inline mini/unroll_test.go:33]: Do we have AddOne, maybe that would be a more interesting test?

File: mini/unroll.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    [(== x '()) (== y '())]
+    [(fresh (xa xd ya yd)
+        (== `(,xa . ,xd) x)
+        (== `(,ya . ,yd) y)
+        (f xa ya)
+        (mapo f xd yd)
+      )]
+  ))
+*/
+func MapO(f func(*ast.SExpr, *ast.SExpr) micro.Goal, x, y *ast.SExpr) micro.Goal {
```
Comment (awalterschulze) [inline mini/unroll.go:65]: Very nice :)

File: mini/unroll.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import (
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+// defines two candidate functions for unrolling
+// in order to check and apply parallel calls
+
+/*
+MemberO checks membership in a list
```
Comment (awalterschulze) [inline mini/unroll.go:12]: Nice :)

File: mini/concurrency_test.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		t.Fatalf("expected len %d, but got len %d instead", numGoals, len(sexprs))
+	}
+	for i, sexpr := range sexprs {
+		if sexpr.String() != fmt.Sprint(i) {
+			t.Fatalf("expected %s, but got %s instead", fmt.Sprint(i), sexpr.String())
+		}
+	}
+}
+
+// and-parallelism is breaking my brain, since in minikanren conj uses bind
```
Comment (awalterschulze) [inline mini/concurrency_test.go:206]: Maybe move this to the pull request description, rather than a comment in the code?

File: mini/concurrency_test.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	"testing"
+	"time"
+
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+// remember, concurrency != parallelism
+
+// setting up exclusive-or for a very simple or-parallelism demo
```
Comment (awalterschulze) [inline mini/concurrency_test.go:14]: Rewrite this comment to just mention exclusive or

File: mini/concurrency_test.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/awalterschulze/gominikanren/micro"
+	"github.com/awalterschulze/gominikanren/sexpr/ast"
+)
+
+// remember, concurrency != parallelism
```
Comment (awalterschulze) [inline mini/concurrency_test.go:12]: Move this comment down to `concurrentDisjExcl`

File: mini/concurrency_test.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	}
+}
+
+// note: this is a naive implementation with runaway goroutines;
+// in a proper implementation the goroutines are taken from a pool
+// and properly cancelled when returning early
+func concurrentDisjExcl(g1, g2 micro.Goal) micro.Goal {
+	return func() micro.GoalFn {
+		return func(s *micro.State) micro.StreamOfStates {
+			c1 := make(chan micro.StreamOfStates, 1)
```
Comment (awalterschulze) [inline mini/concurrency_test.go:51]: Do they have to be buffered?  If so add a comment explaining

File: mini/concurrency_test.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+			c2 := make(chan micro.StreamOfStates, 1)
+			go func() {
+				c1 <- g1()(s)
+			}()
+			go func() {
+				c2 <- g2()(s)
+			}()
+			var g1s, g2s micro.StreamOfStates
+			for i := 0; i < 2; i++ {
```
Comment (awalterschulze) [inline mini/concurrency_test.go:60]: deriveJoin can do this, if you also call `close` inside the go routines.
See here for example https://github.com/awalterschulze/goderive/blob/master/test/normal/join_test.go#L235

File: mini/concurrency_test.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+	}
+}
+
+// illustrating or-parallelism for exclusive-or, where the first answer we find
+// is the one we return.
+func TestOrParallelismExcl(t *testing.T) {
+	// use Zzz to delay evaluation of time.Sleep
+	heavy := func(x *ast.SExpr) micro.Goal {
+		return micro.Zzz(func() micro.GoalFn {
+			time.Sleep(5 * time.Second)
```
Comment (awalterschulze) [inline mini/concurrency_test.go:85]: So if StreamOfStates was a chan instead of a stream, then we could more easily test this and replace the sleep with a channel that only gets triggered once we receive and answer.  Now I can't think of a guaranteed work around :(

File: micro/disj.go [hunk 10]
```diff
@@ -11,15 +11,7 @@ scheme code:
...
-func DisjointO(gs ...Goal) Goal {
-	if len(gs) == 0 {
-		return FailureO
-	}
-	if len(gs) == 1 {
-		return gs[0]
-	}
-	g1 := gs[0]
-	g2 := DisjointO(gs[1:]...)
+func DisjointO(g1, g2 Goal) Goal {
```
Comment (awalterschulze) [inline micro/disj.go:14]: Why make the new function concurrentDisjPlus have elipses, but remove them here?
`func concurrentDisjPlus(gs ...micro.Goal) micro.Goa;`

Comment (awalterschulze) [reply micro/disj.go:14]: Whichever way we go, lets keep the symmetry

File: mini/concurrency_test.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+
+	return func() micro.GoalFn {
+		return func(s *micro.State) micro.StreamOfStates {
+			c1 := make(chan micro.StreamOfStates, 1)
+			go func() {
+				c1 <- g1()(s)
+			}()
+			g2s := g2()(s)
+			var g1s micro.StreamOfStates
+			select {
```
Comment (awalterschulze) [inline mini/concurrency_test.go:132]: Why select on a single channel?

File: mini/concurrency_test.go [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+	return concurrentDisjPlus(conj...)
+}
+
+// this one is more general and useable, but hard to find a good example of it
+// providing value in a simple testcase
+// test with:
+// go test ./... -v -run TestOrParallelism -count=1
+// in order to see the time taken by both
+func TestOrParallelism(t *testing.T) {
+	//numGoals := 50000
```
Comment (awalterschulze) [inline mini/concurrency_test.go:155]: Lets remove the comment :)

File: mini/concurrency_test.go [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+		}
+		return goals
+	}
+	start := time.Now()
+	sexprs := micro.Run(-1, func(q *ast.SExpr) micro.Goal {
+		return concurrentConde(
+			generator(q)...,
+		)
+	})
+	took := time.Now().Sub(start)
```
Comment (awalterschulze) [inline mini/concurrency_test.go:178]: Maybe rewrite these as benchmarks, then you don't have to manually do the timings?

**Review by awalterschulze — COMMENTED** [review]
I am a bit nit picky here sorry, but it is because I think it is good and pretty close to merging.

File: mini/unroll_test.go (hunk 1 above)
Comment (deosjr) [reply mini/unroll_test.go:44]: We don't have any arithmetic yet. Numbers are a pretty involved separate discussion.
I'm avoiding the implementation of fib() for the same reason ;)

File: mini/concurrency_test.go (hunk 4 above)
Comment (deosjr) [reply mini/concurrency_test.go]: Yeah this is part of what needs to be removed after we've discussed some more, I think.
Included because I wanted to show you my thought process on future direction

File: micro/disj.go (hunk 10 above)
Comment (deosjr) [reply micro/disj.go:14]: Theres a difference between what the paper calls `disj` and `conj`, which only take two goals, and macros `disj+` and `conj+` which extend these to arbitrary number of goals.
You had a DisjointO and ConjunctionO, which were somewhere in between.
They were variadic but did not include an automatic Zzz wrapper like in the paper.
I fixed DisjointO and ConjunctionO (please lets rename these) to match the elementary `disj` and `conj`.

File: mini/concurrency_test.go (hunk 11 above)
Comment (deosjr) [reply mini/concurrency_test.go:132]: artifact of select on 2, should be removed

File: mini/concurrency_test.go (hunk 13 above)
Comment (deosjr) [reply mini/concurrency_test.go:178]: Yeah I think this whole test needs to be redone, but it illustrated where I was to you, right? :)
Think it served its purpose and can be removed/replaced.

File: micro/disj.go [hunk 14]
```diff
@@ -46,17 +46,17 @@ scheme code:
 		)
 	)
 */
-func Mplus(g1, g2 StreamOfStates) StreamOfStates {
-	if g1 == nil {
-		return g2
+func Mplus(s1, s2 StreamOfStates) StreamOfStates {
```
Comment (awalterschulze) [inline micro/disj.go:49]: Nice cleanup

File: micro/exts.go [hunk 15]
```diff
@@ -6,7 +6,7 @@ import (
 
 /*
 exts either extends a substitution s with an association between the variable x and the value v ,
-or it produces (ok = false)s
+or it produces (ok = false)
```
Comment (awalterschulze) [inline micro/exts.go:9]: Thanks :)

File: mini/einstein_test.go [hunk 16]
```diff
@@ -0,0 +1,194 @@
+package mini
```
Comment (awalterschulze) [inline mini/einstein_test.go:1]: If this doesn't use concurrency, I think this can be a separate PR?

File: mini/einstein_test.go [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+		)
+	}
+
+	// NOTE: these should be anonymous vars, but we cannot make those atm without using call/fresh
+	// I'm creating vars with unique random names (from unix time) instead, for now
+	// somehow ast.NewVariable can introduce vars with scope: if you change one of the var names in
+	// street to duplicate another, you will see them substituted in variable bindings in state
+	// I don't think miniKanren _should_ work like that..
+	a := func() *ast.SExpr {
+		return ast.NewVariable(time.Now().String())
```
Comment (awalterschulze) [inline mini/einstein_test.go:85]: There is a single non pure functional part of mini kanren, which is using ref in scheme.
This makes it possible to create unique variables, if I recall correctly
It might have something to do with what you do here.

**Review by awalterschulze — COMMENTED** [review]
Maybe we can continue this review on Tuesday in the call.
It all looks very interesting :)

File: mini/einstein_test.go [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+The German smokes Prince.
+The Norwegian lives next to the blue house.
+The Blend smoker has a neighbor who drinks water.
+
+The question to be answered is: Who keeps fish?
+Answer adapted from https://gist.github.com/jgilchrist/1b85b04f4395057972dd
+*/
+
+// takes almost a minute (52s) to run atm on my machine with MemberO
+// takes about 5s to run with MemberO unrolled, _without any concurrency support_
```
Comment (deosjr) [inline mini/einstein_test.go:37]: For reference, the minikanren-in-scheme version runs in 0.35s or so, without unrolling.
We have some optimisation to do :)

Comment (deosjr) [reply mini/einstein_test.go:37]: One more note: if I change the substitution map from string->sexpr to int64->sexpr, we shave off another 10% running time or so. It also brings us more inline with existing ukanren/minikanren implementations.
I think that is one more thing we need can do (separate PR), but I run into issues with the parser doing it.
In order to run it locally I had to delete all other tests, which is not ideal ;)

Comment (awalterschulze) [reply mini/einstein_test.go:37]: If you go that route, go also try uint64.  I recall it being faster for map lookup, but that was 6 years ago.

Although a 10% shaving is far from the two orders of magnitude that we need.  So might be worth looking for some bigger wins first, but that is your choice.  I just find the strings easier to read.

Comment (deosjr) [reply mini/einstein_test.go:37]: tbh the main argument for doing it early is that we can compare with other implementations.
This variable naming and conflation of concepts with the parser has confused a bunch of times already.

I am comparing with fast-minikanren btw, so that is optimised in a bunch of ways. 
Not sure how naive minikanren-in-chezscheme would perform.
I've been digging deep into the fast-minikanren implementation and almost have a grasp of all the moving parts;
there's a lot of things we can do with checking the scope of vars. But again, we need vars to be cleaned up first.

I kinda need you for the parser bit, since I'm not sure on the details there and what might break.
Haven't ever _used_ the parser either, just writting gominikanren funcs by hand in Go :)

Comment (awalterschulze) [reply mini/einstein_test.go:37]: Cool sounds good.  Lets also discuss the parser on Tuesday :)

**Comment by deosjr** [comment]
Closing this PR and splitting the good parts into a few new ones
